Quest-3/Part-1/main.py

```diff
@@ -0,0 +1,71 @@
+def main(fileName):
+    lines = readFile(fileName)
+
+    schemEngine = [list(line.rstrip("\n")) for line in lines]
+    symbols = set(char for line in schemEngine for char in line if char != '.' and not char.isdigit())
+
+    allNumbers = findNumbers(schemEngine)
+    trueNumbers = findSymbolsAroundNumbers(schemEngine, allNumbers, symbols)
+
+    total = sum(trueNumbers)
+    print(total)
+
+def readFile(name):
+    with open(file=f"Quest-3/{name}", mode="r", encoding="utf-8") as f:
+        return f.readlines()
+
+def findNumbers(matrix):
+    numbers = []
+    visited = set()
+    rows = len(matrix)
+    cols = len(matrix[0])
+
+    def explore(row, col):
+        num = []
+        num.append({"digit": matrix[row][col], "coordinates": [row, col]})
+        # Explore horizontally
+        for c in range(col + 1, cols):
+            if matrix[row][c].isdigit() and (row, c) not in visited:
+                num.append({"digit": matrix[row][c], "coordinates": [row, c]})
+                visited.add((row, c))
+            else:
+                break
+        return num
+
+    for row in range(rows):
+        for col in range(cols):
+            if matrix[row][col].isdigit() and (row, col) not in visited:
+                num = explore(row, col)
+                digit = [el["digit"] for el in num]
+                coordinates = [el["coordinates"] for el in num]
+
+                numbers.append([int("".join(digit)), coordinates])
+
+    return numbers
+
+def findSymbolsAroundNumbers(matrix, numbers, symbols):
+    results = []
+
+    for numberKey, coordinates in numbers:
+        foundSymbol = False
+        for coord in coordinates:
+            row, col = coord
+            for dr, dc in [(0, 1), (0, -1), (1, 0), (-1, 0), (1, 1), (1, -1), (-1, 1), (-1, -1)]:
+                newRow, newCol = row + dr, col + dc
+                if (newRow, newCol) in coordinates:
+                    continue  # Skip a cell if it is part of the same number
+                if 0 <= newRow < len(matrix) and 0 <= newCol < len(matrix[0]) and matrix[newRow][newCol] in symbols:
+                    foundSymbol = True
```

Comment:
instead of a flag `foundSymbol` and multiple breaks, you can extract this loop in a function and return True from inside of the loop. This is usually much better. Flags of any sort suck.